Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,363 @@
+
+        private readonly ConcurrentBag<MuTuple<Func<IChannel, object>, Func<IChannel, Task>>> GlobalChannelListeners = new ConcurrentBag<MuTuple<Func<IChannel, object>, Func<IChannel, Task>>>();
+        private readonly ConcurrentDictionary<int, IList<MuTuple<Func<IChannel, object>, Func<IChannel, Task>>>> SingleChannelListeners = new ConcurrentDictionary<int, IList<MuTuple<Func<IChannel, object>, Func<IChannel, Task>>>>();
+        private readonly ConcurrentDictionary<int, IList<MuTuple<Func<IChannel, object>, Func<IChannel, Task>, object>>> ChannelListeners = new ConcurrentDictionary<int, IList<MuTuple<Func<IChannel, object>, Func<IChannel, Task>, object>>>();
+
+        public async sealed override void Init()
+        {
+            base.Init();
+            bool success = true;
+            foreach (var user in await Server.GetOnlineUsers()) {
+                success = UserListeners.TryAdd(user.SessionId, new List<MuTuple<Func<IUser, object>, Func<IUser, Task>, object>>()) ? success : false;
+            }
+            foreach (var channel in await Server.GetChannels()) {
+                success = ChannelListeners.TryAdd(channel.ChannelId, new List<MuTuple<Func<IChannel, object>, Func<IChannel, Task>, object>>()) ? success : false;
+            }
+
+            if (!success) {
+                throw new Exception("Init of FancyPluginBase failed due to non-empty listener-lists or duplicte IDs. " +
+                    "This should actually not happen - we are so screwed!!!");
+            }
+        }
+
+        private void AddGlobalListener<T>(ConcurrentBag<MuTuple<Func<T, object>, Func<T, Task>>> globalListeners,
+            ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners,
+            Func<T, object> mapper, Func<T, Task> listener) where T : IThing
+        {
+            globalListeners.Add(MuTuple.Create(mapper, listener));
+            foreach (var key in listeners.Keys) {
+                listeners.AddOrUpdate(key,
+                    i =>
+                    {
+                        var l = new List<MuTuple<Func<T, object>, Func<T, Task>, object>>();
+                        l.Add(MuTuple.Create(mapper, listener, (object)null));
+                        return l;
+                    },
+                    (i, l) =>
+                    {
+                        l.Add(MuTuple.Create(mapper, listener, (object)null));
+                        return l;
+                    }
+                );
+            }
+        }
+
+        private void AddSingleListener<T>(ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>>>> singleListeners,
+            ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners,
+            Func<T, object> mapper, Func<T, Task> listener,
+            int uniqueId, bool permanent) where T : IThing
+        {
+            if (permanent) {
+                singleListeners.AddOrUpdate(uniqueId,
```

Comment:
Looks like your `uniqueId` is the session id. Problem: Reconnecting might give you the same session id again, but it probably won't. Oh, and you might get a session id that someone else (who has disconnected) had. So this won't ever work as intended. (unless I'm blind)